pkg/engine/engine.go

```diff
@@ -200,6 +204,38 @@ func execRunnable(ctx context.Context, log logr.Logger, r Runnable) error {
 	return nil
 }
 
+// Reset re-initializes engine and deletes the remaining objects
 func (eng *Eng) Reset(ctx context.Context) error {
+	eng.log.Info("Reset Engine")
+
+	if eng.cleanup == nil || !eng.cleanup.Enabled {
+		return nil
+	}
+
+	eng.log.Info("Cleaning up objects")
+	ctx, cancel := context.WithTimeout(ctx, eng.cleanup.Timeout)
+	defer cancel()
+
+	return eng.DeleteAllObjects(ctx)
+}
+
+// DeleteAllObjects deletes all objects
+func (eng *Eng) DeleteAllObjects(ctx context.Context) error {
+	deletePolicy := metav1.DeletePropagationBackground
+	deletions := metav1.DeleteOptions{
+		PropagationPolicy: &deletePolicy,
+	}
+
+	for _, objInfo := range eng.objMap {
+		ns := objInfo.Namespace
+		for _, name := range objInfo.Names {
+			err := eng.dynamicClient.Resource(objInfo.GVR).Namespace(ns).Delete(ctx, name, deletions)
+			if err != nil {
+				return err
```

Comment:
Would it return an error if the object has been already deleted by the test or expired naturally?

Reply:
Is this correct way to add a `DeleteObj` task? The task id is `job` in the `SubmitObj` task. 
```
- id: delete
  type: DeleteObj
  params:
    refTaskId: job
  refTaskId: job
  ```
  
  
 
 
 When I ran the test (without --cleanup) and got the following error.
```
E0516 09:56:44.038701   53311 engine.go:200] "Task failed" err="the server could not find the requested resource" id="DeleteObj/delete"
I0516 09:56:44.038712   53311 engine.go:209] "Reset Engine"
E0516 09:56:44.038718   53311 main.go:96] the server could not find the requested resource
```

Reply:
Also, I was wondering why we need to define refTaskId twice in `DeleteObj`.

Reply:
No, we don't need refTaskId twice.

Reply:
I removed the second `refTaskId` and still got the same error.  Do we have a test with a `DeleteObj` task other than the unit test? Can you try it by yourself? 
```
 ./bin/knavigator --tasks resources/tests/k8s/test-job.yml
I0516 10:15:41.155408   53935 k8s_config.go:42] "Using external kubeconfig"
I0516 10:15:41.158578   53935 main.go:84] "Starting test" name="test-k8s-job"
I0516 10:15:41.158590   53935 engine.go:103] "Creating task" name="SubmitObj" id="job"
I0516 10:15:41.160156   53935 engine.go:197] "Starting task" id="SubmitObj/job"
I0516 10:15:41.182176   53935 engine.go:203] "Task completed" id="SubmitObj/job" duration="22.011666ms"
I0516 10:15:41.182188   53935 engine.go:103] "Creating task" name="DeleteObj" id="delete"
I0516 10:15:41.182224   53935 engine.go:197] "Starting task" id="DeleteObj/delete"
E0516 10:15:41.184724   53935 engine.go:200] "Task failed" err="the server could not find the requested resource" id="DeleteObj/delete"
I0516 10:15:41.184735   53935 engine.go:209] "Reset Engine"
E0516 10:15:41.184741   53935 main.go:96] the server could not find the requested resource
```

Reply:
indeed, we never tested it. Let me check

Reply:
A bug indeed. Fixed in https://github.com/NVIDIA/knavigator/pull/41

Reply:
I tested `--cleanup` with the `DeleteObj` task. It's working. No error.  The deleted object should be removed from ObjInfoMap?
```
./bin/knavigator --tasks resources/tests/k8s/test-job.yml  --cleanup
I0517 10:24:41.489049   93711 k8s_config.go:42] "Using external kubeconfig"
I0517 10:24:41.494671   93711 main.go:84] "Starting test" name="test-k8s-job"
I0517 10:24:41.494690   93711 engine.go:111] "Creating task" name="RegisterObj" id="register"
I0517 10:24:41.495255   93711 engine.go:244] "Starting task" id="RegisterObj/register"
I0517 10:24:41.495266   93711 engine.go:250] "Task completed" id="RegisterObj/register" duration="1.708µs"
I0517 10:24:41.495271   93711 engine.go:111] "Creating task" name="SubmitObj" id="job"
I0517 10:24:41.495349   93711 engine.go:244] "Starting task" id="SubmitObj/job"
I0517 10:24:41.512305   93711 engine.go:250] "Task completed" id="SubmitObj/job" duration="16.949084ms"
I0517 10:24:41.512316   93711 engine.go:111] "Creating task" name="CheckPod" id="status"
I0517 10:24:41.512363   93711 engine.go:244] "Starting task" id="CheckPod/status"
I0517 10:24:41.512369   93711 check_pod_task.go:158] "Create pod informer" #pod=2 timeout="5s"
I0517 10:24:41.517275   93711 check_pod_task.go:256] "Accounted for all pods"
I0517 10:24:41.517300   93711 engine.go:250] "Task completed" id="CheckPod/status" duration="4.932042ms"
I0517 10:24:41.517317   93711 engine.go:111] "Creating task" name="DeleteObj" id="delete"
I0517 10:24:41.517385   93711 engine.go:244] "Starting task" id="DeleteObj/delete"
W0517 10:24:41.529210   93711 warnings.go:70] child pods are preserved by default when jobs are deleted; set propagationPolicy=Background to remove them or set propagationPolicy=Orphan to suppress this warning
I0517 10:24:41.529230   93711 engine.go:250] "Task completed" id="DeleteObj/delete" duration="11.837917ms"
I0517 10:24:41.529235   93711 engine.go:256] "Reset Engine"
I0517 10:24:41.529237   93711 engine.go:262] "Cleaning up objects"
I0517 10:24:41.533397   93711 engine.go:286] "Deleted all objects"
```